wp/lib/bap_wp/src/precondition.ml

```diff
@@ -396,62 +399,84 @@ let subst_fun_outputs ?tid_name:(tid_name = "") ~inputs:(inputs : Var.t list)
       Env.add_call_pred env sub_to) in
   Constr.substitute post subs_from subs_to, env
 
-let input_regs (arch : Arch.t) : Var.t list =
-  match arch with
-  | `x86_64 ->
-    let open X86_cpu.AMD64 in
-    (* r.(0) and r.(1) refer to registers R8 and R9 respectively.
-       Arguments are placed on the stack when they have a higher count than the
-       number of registers. We currently do not handle mem as an input because it
-       causes Z3 to slow down during evaluation. *)
-    info "[mem] is not included as an input to the function call.%!";
-    [rdi; rsi; rdx; rcx; r.(0); r.(1)]
-  | `x86 ->
-    warning "In 32-bit x86, arguments are passed through the stack.%!";
-    []
-  | #Arch.arm ->
-    let open ARM.CPU in
-    [r0; r1; r2; r3]
-  | a ->
-    warning "input_regs: Input registers have not been \
-             implemented for %s." (Arch.to_string a);
-    []
-
-let caller_saved_regs (arch : Arch.t) : Var.t list =
-  match arch with
-  | `x86_64 ->
-    let open X86_cpu.AMD64 in
-    (* Obtains registers r8 - r11 from X86_cpu.AMD64.r. *)
-    let r = Array.to_list (Array.sub r ~pos:0 ~len:4) in
-    [rax; rcx; rdx; rsi; rdi] @ r
-  | `x86 ->
-    let open X86_cpu.IA32 in
-    [rax; rcx; rdx]
-  | #Arch.arm ->
-    let open ARM.CPU in
-    [r0; r1; r2; r3; r12]
-  | a ->
-    warning "caller_saved_regs: Caller-saved registers have not \
-             been implemented for %s." (Arch.to_string a);
-    []
-
-let callee_saved_regs (arch : Arch.t) : Var.t list =
-  match arch with
-  | `x86_64 ->
-    let open X86_cpu.AMD64 in
-    (* Obtains registers r12 - r15 from X86_cpu.AMD64.r. *)
-    let r = Array.to_list (Array.sub r ~pos:4 ~len:4) in
-    [rbx; rsp; rbp] @ r
-  | `x86 ->
-    let open X86_cpu.IA32 in
-    [rbx; rdi; rsi; rsp; rbp]
-  | #Arch.arm ->
-    let open ARM.CPU in
-    [r4; r5; r6; r7; r8; r9; r10; r11]
-  | a ->
-    warning "callee_saved_regs: Callee-saved registers have not \
-             been implemented for %s." (Arch.to_string a);
-    []
+(* FIXME: use built-in BAP roles? *)
+let input_regs (arch : Theory.target) : Var.t list =
+  if (Theory.Target.matches arch "amd64") then
```

Comment:
Having magic strings all over these functions makes me a bit nervous.  Can we define the targets outside of the function and use those variables as identifiers instead? We are matching on the strings explicitly in many places in `Environment.ml` and `Precondition.ml`.

```
let amd64 = "amd64"

...

if (Theory.Target.matches arch amd64) then ...
```

Reply:
One might argue that these strings are not really magic at all, since they represent exactly what they mean.

I agree that that bap likes to use `amd64` rather than `x86_64` is somewhat non-obvious, but I'm not sure how to resolve this.

We can hoist this out into a single let-binding, if you like.

Reply:
On reflection, this would probably be cleaner with `is_amd64`, `is_arm` etc. functions

Reply:
That sounds good! We already have an `is_x86` so maybe it would be a good idea to have for `arm` etc.

I just want to avoid the situation where someone writes something like `Theory.Target.matches arch "amd644"` and we don't catch it at compile time.

Reply:
Reasonable.